File: engine/core/FlyingCamera.cpp
#include "FlyingCamera.h"

#include "Game.h"


void FlyingCamera::OnInit() {
	m_mouseMoveHandle = game()->input()->MouseMove.AddRaw(this, &FlyingCamera::m_OnMouseMove);
	UpdateProjectionMatrix(/*game()->window()*/);

	game()->hotkeys()->RegisterHotkey(Keys::O);
}

void FlyingCamera::OnDestroy() {
	game()->input()->MouseMove.Remove(m_mouseMoveHandle);
	game()->hotkeys()->UnregisterHotkey(Keys::O);
}

void FlyingCamera::OnUpdate() {
	if (!IsAttached())
		return;

	if (game()->hotkeys()->Is(Keys::O, KeyState::Press))
		orthographic(!orthographic());

	if (m_updateRotation) {
		auto rot = localRotation();
		rot += m_rotationDelta;
		localRotation(rot);
		m_rotationDelta = Vector3::Zero;
		m_updateRotation = false;
	}

	auto rot = localRotation();
	auto rotator = Matrix::CreateFromYawPitchRoll(rot.y, rot.x, 0);

	auto input = game()->input();
	if (input->IsKeyDown(Keys::RightButton)) {

		auto axis = Vector3::Zero;
		if (input->IsKeyDown(Keys::W)) axis += Vector3::Right;
		if (input->IsKeyDown(Keys::S)) axis += Vector3::Left;
		if (input->IsKeyDown(Keys::A)) axis += Vector3::Forward;
		if (input->IsKeyDown(Keys::D)) axis += Vector3::Backward;
		if (input->IsKeyDown(Keys::Space)) axis += Vector3::Up;
		if (input->IsKeyDown(Keys::LeftShift)) axis += Vector3::Down;
		axis.Normalize();

		auto direction = rotator.Forward() * axis.x + Vector3::Up * axis.y + rotator.Right() * axis.z;
		direction.Normalize();

		auto newPos = localPosition() + direction * m_speed * game()->deltaTime();
		localPosition(newPos);
	}

	//auto m = GetWorldMatrix();
	auto newMatrix = Matrix::CreateLookAt(worldPosition(), worldPosition() + rotator.Forward(), rotator.Up());
	viewMatrix(newMatrix);
	UpdateProjectionMatrix(/*game()->window()*/);

	if (printTransform)
		game()->SendGameMessage(std::to_string(actor()->Id()) + " tr");
}

void FlyingCamera::m_OnMouseMove(const InputDevice::MouseMoveArgs& args) {
	if (!IsAttached())
		return;

	if (!game()->input()->IsKeyDown(Keys::RightButton))
		return;

	m_rotationDelta.y -= args.Offset.x * 0.003f * m_mouseSense;
	m_rotationDelta.x -= args.Offset.y * 0.003f * m_mouseSense;
	m_updateRotation = true;

	if (args.WheelDelta > 0)
		m_speed += 1;
	if (args.WheelDelta < 0)
		m_speed -= 1;
}

void FlyingCamera::RecieveGameMessage(const std::string& msg) {
	if (msg == "camera.print.tr")
		printTransform = !printTransform;
}

File: engine/mono/Scripts/FlyingCamera.cs
using System;

using Engine;


[GUID("d4806d9b-d18b-4936-873c-a356a8f50288")]
public class FlyingCamera : CSComponent {

    [Open] private CameraComponent m_camera;

    public float rotationSense = 0.2f;

    private bool m_updateRotation = false;
    private Vector3 m_rotationDelta;
    private float m_speed = 100.0f;

    public override void OnInit() {
        if (m_camera == null)
            throw new Exception($"{GetType().FullName}: Not all fields are valid");

        Input.OnMouseMove += m_OnMouseMove;
        Input.OnWheelMove += m_OnWheelMove;
    }

    public override void OnDestroy() {
        Input.OnMouseMove -= m_OnMouseMove;
        Input.OnWheelMove -= m_OnWheelMove;
    }

    public override void OnStart() {
        m_camera.UpdateProjMatrix();
        // m_camera.Attach();
    }

    public override void OnFixedUpdate() {
        if (!m_camera.IsAttached)
            return;

        if (m_updateRotation) {
            var lrot = actor.localRotation;
            lrot += m_rotationDelta;
            actor.localRotation = lrot;
            m_rotationDelta = Vector3.Zero;
            m_updateRotation = false;
        }

        var axis = Vector3.Zero;
        if (Input.GetButton(Key.W)) axis += Vector3.Forward;
        if (Input.GetButton(Key.S)) axis += Vector3.Backward;
        if (Input.GetButton(Key.A)) axis += Vector3.Left;
        if (Input.GetButton(Key.D)) axis += Vector3.Right;
        if (Input.GetButton(Key.Space)) axis += Vector3.Up;
        if (Input.GetButton(Key.LeftShift)) axis += Vector3.Down;
        axis = axis.Normalized();

        var rot = actor.localRotation;
        var rotator = Quaternion.CreateFromYawPitchRoll(rot.Y, rot.X, 0);
        var rotatorForward = Vector3.Forward.Rotate(rotator);
        var rotatorRight = Vector3.Right.Rotate(rotator);
        var rotatorUp = Vector3.Up.Rotate(rotator);

        var direction = rotatorForward * -axis.Z + Vector3.Up * axis.Y + rotatorRight * axis.X;
        direction = direction.Normalized();

        var newPos = actor.localPosition + direction * m_speed * Game.DeltaFixedTime;
        actor.localPosition = newPos;

        var newMatrix = Matrix4x4.CreateLookAt(actor.worldPosition, actor.worldPosition + rotatorForward, rotatorUp);
        m_camera.viewMatrix = newMatrix;
    }

    private void m_OnMouseMove() {
        if (!m_camera.IsAttached)
            return;

        m_rotationDelta.Y -= Input.MouseDelta.X * rotationSense;
        m_rotationDelta.X -= Input.MouseDelta.Y * rotationSense;
        m_updateRotation = true;
    }

    public void m_OnWheelMove() {
        if (!m_camera.IsAttached)
            return;

        if (Input.WheelDelta > 0)
            m_speed += 1;
        if (Input.WheelDelta < 0)
            m_speed -= 1;
    }

}